tools/monitoring/alert/report.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alert
+
+import (
+	"database/sql"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/knative/test-infra/tools/monitoring/config"
+	"github.com/knative/test-infra/tools/monitoring/log_parser"
+)
+
+const emailTemplate = `In the past %v, 
+The following error pattern reached alerting threshold:
+%s
+
+# occurrences: %v
+jobs affected: %v
+PRs affected: %v
+
+Hint for diagnose & recovery: %s
+
+Error Logs:
+%v
+`
+
+// report stores list of error logs, together with sets of jobs/PRs in those logs
+type report struct {
+	logs []log_parser.ErrorLog
+	jobs []string
+	prs  []int
+}
+
+type mailContent struct {
+	report
+	errorPattern string
+	hint         string
+	window       time.Duration
+}
+
+func (c mailContent) body() string {
```

Comment:
Do you have a sample email body generated from this function? 

From the code, I think `c.logs` prints out the whole ErrorLog structure i.e. Pattern, Msg, JobName, PRNumber, BuildLogURL, TimeStamp. Maybe you meant to print out a subset of those fields?

Reply:
No, I don't

I meant to print out the whole structure. I don't think any field is useless to include here.
It might look a bit messy. In the future we can make it a table under html formatting, yet I don't think that is an immediate need